src/actions/cart-manager.ts

```diff
@@ -0,0 +1,92 @@
+import {retrieveSpecificProduct} from './DAL/product-db-actions';
+
+export const mainCart: { id: string, users: { userId: string, amount: number }[] }[] = [];
+
+export const addToCart = async (ip: string, productId: string, desiredAmount: number) => {
+    const product = await retrieveSpecificProduct(productId);
+    let availableAmount;
+    if (!product)
+        throw new Error('Item not found!');
+    let productInMainCart = findProductInMainCart(productId);
+    if (productInMainCart) {
+        availableAmount = product.quantity - getAmountOfItemInMainCart(productInMainCart);
+    } else {
+        availableAmount = product.quantity;
+        productInMainCart = {id: productId, users: []};
+        mainCart.push(productInMainCart);
+    }
+    if (desiredAmount <= availableAmount)
+        updateProductInMainCart(ip, desiredAmount, productInMainCart);
+    else
+        throw new Error('Quantity not enough');
+    return availableAmount - desiredAmount;
+};
+
+export const removeFromCart = async (userId: string, productId: string) => {
+    const product = await retrieveSpecificProduct(productId);
+    if (!product)
+        throw new Error('Item not found!');
+    const productInMainCart = findProductInMainCart(productId);
+    if (productInMainCart) {
+        removeUserFromItemInMainCart(userId, productInMainCart);
+        return product.quantity - getAmountOfItemInMainCart(productInMainCart);
+    } else
+        throw new Error('Item not found!');
+};
+
+const findProductInMainCart: (id: string) => any = (id: string) => {
+    let foundItem;
+    mainCart.forEach(cartItem => {
+        if (cartItem.id === id)
+            foundItem = cartItem;
+    });
+    return foundItem ? foundItem : null;
+};
+
+const getAmountOfItemInMainCart = (cartItem: { id: string, users: { userId: string, amount: number }[] }) => {
+    let amountInCarts = 0;
+    cartItem.users.forEach(user => amountInCarts += user.amount);
+    return amountInCarts;
+};
+
+const updateProductInMainCart = (userId: string, amount: number, cartItem: { id: string, users: { userId: string, amount: number }[] }) => {
```

Comment:
create type for cartItem

Reply:
Now the mainCart is a Map of Maps. it's much simpler